02-express-tutorial/app.js

```diff
@@ -1 +1,97 @@
-console.log('Express Tutorial')
+const express = require('express');
+const app = express();
+const { products } = require("./data")
+
+
+/**
+ * Create HTTP server
+ */
+// const http = require('http')
+const port = 3000;
+// app.set('port', port)
+// const server = http.createServer(app)
+// server.listen(port, () => {
+//     console.log(`app listening at http://localhost:${port}`)
+// })
+
+app.use(express.static("./public"))
+
+// First test endpoint
+app.get('/api/v1/test', (request, response) => {
+    response.json({
+        message: 'My test API endpoint!'
+    })
+})
+
+// Get all products
+app.get('/api/v1/products', (request, response) => {
+    response.json(products)
+})
+
+// Get product by Id
+app.get('/api/v1/products/:productId', (request, response) => {
+    
+    const productId = parseInt(request.params.productId)
+    const product = products.find( p => p.id === productId )
+    
+    if( !product ) {
+        response.status(400).json({
+            message: 'Product not found!'
+        })
+    }
+
+    response.json(product)
+})
+
+// Search product query
+app.get('/api/v1/query', (request, response) => {
+    const defaultLimit = 5;
+
+    const search = request.query.search && request.query.search.toLocaleLowerCase() || ""
+    const limit = Math.abs( parseInt(request.query.limit) ) || defaultLimit
+    const minPrice = parseFloat(request.query.minPrice) || -1
+    const maxPrice = parseFloat(request.query.maxPrice) || -1
+
+    // Handle error: Always minPrice < maxPrice
+    if( minPrice!==-1 && maxPrice!==-1 && minPrice>maxPrice ) {
+        response.status(400).json({
+            message: "Minimum and maximum price conflict!"
+        })
+    }
+    
+    let resultProducts = []
+    // Check if at least one query param provided to filter the result
+    if( search !== "" || minPrice !== -1 || maxPrice !== -1 ) {
```

Comment:
Instead of combining the logic for all three query parameters in a single block, I recommend that you have separate blocks for each query parameter.  Try to keep things as decoupled from each other whenever you can.